src/main/java/core/reader/TabDelimitedReader.java

```diff
@@ -0,0 +1,203 @@
+      eventType = EventType.valueOf(columns[0]);
+    } catch (Exception e) {
+      validationErr("invalid event type", e, columns);
+    }
+
+    switch (eventType) {
+      case ACCOUNT:
+        accountDetails.add(rowToAccount(columns));
+        break;
+      case TRANSACTION:
+        transactions.add(rowToTransaction(columns));
+        break;
+      default:
+        logicErr(columns);
+    }
+
+  }
+
+  private AccountDetails rowToAccount(String[] columns) {
+    String number = columns[1].trim();
+    String name = columns[2].trim();
+    String increasesOn = columns[3].trim();
+    AccountDetails y = null;
+    try {
+      y = new AccountDetailsImpl(
+              number,
+              name,
+              AccountSide.valueOf(increasesOn));
+    } catch (Exception e) {
+      validationErr("invalid account data", e, columns);
+    }
+    return y;
+  }
+
+  private AccountingTransaction rowToTransaction(String[] columns) {
+    String dateTime = columns[4].trim();
+    String debitAccounts = columns[5].trim();
+    String debitAmounts = columns[6].trim();
+    String creditAccounts = columns[7].trim();
+    String creditAmounts = columns[8].trim();
+
+
+    Set<AccountingEntry> entries = toAccountingEntryset(
+        DEBIT,
+        debitAccounts,
+        debitAmounts);
+    entries.addAll(toAccountingEntryset(
+        CREDIT,
+        creditAccounts,
+        creditAmounts));
+
+    // 2011-12-03T10:15:30+01:00
+    long transactionTime = ZonedDateTime
+        .parse(dateTime, ISO_OFFSET_DATE_TIME)
+        .toInstant()
+        .toEpochMilli();
+
+    AccountingTransaction y = null;
+    try {
+      y = new AccountingTransaction(
+              entries,
+              null,
+              transactionTime);
+    } catch (Exception e) {
+      validationErr("invalid transaction data", e, columns);
+    }
+    return y;
+  }
+
+  private Set<AccountingEntry> toAccountingEntryset(
+      AccountSide accountSide,
+      String accounts,
+      String amounts
+  ) {
+
+    Set<AccountingEntry> ys = new HashSet<>();
+    String[] as = accounts.split(",");
+    String[] bs = amounts.split(",");
+    for (int i = 0; i < as.length; i++) {
+      ys.add(
+          // BigDecimal amount, String accountNumber, AccountSide accountSide
+          new AccountingEntry(
+              new BigDecimal(bs[i]),
+              as[i],
+              accountSide
+          )
+      );
+    }
+
+    return ys;
+
+  }
+
+  // Convenience function for raising validate errors.
+  private String errMessage(String msg, String[] columns) {
+    return msg + " on line with columns: " + Arrays.toString(columns);
+  }
+
+  private void logicErr(String[] columns) {
+    throw new IllegalStateException(
+        errMessage("logic error in code", columns));
+  }
+
+  private void validationErr(String msg, Throwable e, String[] columns) {
+    throw new IllegalStateException(errMessage(msg, columns), e);
+  }
+
+
+  public enum EventType {ACCOUNT, TRANSACTION}
```

Comment:
`EntryType` like the header in the test resource files feels like a more descriptive name. What do you think?

Reply:
I agree.